From 77eb39432190cb6e58dfdb8a4865c578643c4968 Mon Sep 17 00:00:00 2001
From: Jonathan David <jonathan.david@ni.com>
Date: Mon, 21 Jul 2014 14:32:25 -0500
Subject: [PATCH] Added curl_multi_setopt_va() API function

Added API function with `va_list` variable argument parameter in
addition to cURL's function using `...`.

Upstream-Status: Denied

https://curl.se/mail/lib-2017-01/0138.html

This patchset was denied by upstream because they did not agree that our
usecase for these added functions were valid - and we did not have a
defense.

https://curl.se/mail/lib-2017-01/0146.html

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

Reapplied to cURL 8.6.0.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>
---
 lib/multi.c | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/lib/multi.c b/lib/multi.c
index 0926b0d..d78f8b2 100644
--- a/lib/multi.c
+++ b/lib/multi.c
@@ -3316,8 +3316,20 @@ static CURLMcode multi_socket(struct Curl_multi *multi,
 CURLMcode curl_multi_setopt(struct Curl_multi *multi,
                             CURLMoption option, ...)
 {
-  CURLMcode res = CURLM_OK;
+  CURLMcode res;
   va_list param;
+  va_start(param, option);
+
+  res = curl_multi_setopt_va(multi, option, param);
+  va_end(param);
+  return res;
+}
+
+/* NI-specific function, with va_list param, instead of extended args (...) */
+CURLMcode curl_multi_setopt_va(struct Curl_multi *multi,
+                               CURLMoption option, va_list param)
+{
+  CURLMcode res = CURLM_OK;
   unsigned long uarg;
 
   if(!GOOD_MULTI_HANDLE(multi))
@@ -3326,8 +3338,6 @@ CURLMcode curl_multi_setopt(struct Curl_multi *multi,
   if(multi->in_callback)
     return CURLM_RECURSIVE_API_CALL;
 
-  va_start(param, option);
-
   switch(option) {
   case CURLMOPT_SOCKETFUNCTION:
     multi->socket_cb = va_arg(param, curl_socket_callback);
